modules/host/network.go

```diff
@@ -174,6 +161,20 @@ func (h *Host) threadedListen() {
 	}
 	defer h.tg.Done()
 
+	// Automatically close the listener when h.tg.Stop() is called.
+	err = h.tg.Add()
+	if err != nil {
+		return
+	}
+	go func() {
+		defer h.tg.Done()
+		<-h.tg.StopChan()
+		err = h.listener.Close()
+		if err != nil {
+			h.log.Println("Closing the listener failed:", err)
+		}
+	}()
+
```

Comment:
I was thinking you could move to starting up the listener into this thread, but really you want to make sure that the `host` object is not returned until the listener is running. Currently the listener is spawned in a thread that will block the return of the host object, and that's the way it should stay.

This code definitely belongs next to the listener, because the job of this code is to close the listener, not to close `threadedListen`. `threadedListen` is a separate entity that chose to close based on the uptime of the listener. If you don't like that then you should be making it's shutdown more explicitly depend on the `StopChan` instead of on the death of the listener, but I don't think that's necessary.

NACK on this code change.